src/cryptonote_core/service_node_voting.cpp

```diff
@@ -70,16 +72,31 @@ namespace service_nodes
     return result;
   }
 
-  static crypto::hash make_deregister_vote_hash(uint64_t block_height, uint32_t service_node_index)
+  static crypto::hash make_deregister_vote_hash(uint32_t version, uint64_t block_height, uint32_t service_node_index, uint32_t vote_type)
   {
-    const int buf_size = sizeof(block_height) + sizeof(service_node_index);
-    char buf[buf_size];
+    crypto::hash result;
+    if (version <= quorum_vote_t::version_0_infinite_staking)
+    {
+      constexpr int buf_size = sizeof(block_height) + sizeof(service_node_index);
+      char buf[buf_size];
 
-    memcpy(buf, reinterpret_cast<void *>(&block_height), sizeof(block_height));
-    memcpy(buf + sizeof(block_height), reinterpret_cast<void *>(&service_node_index), sizeof(service_node_index));
+      memcpy(buf, reinterpret_cast<void *>(&block_height), sizeof(block_height));
+      memcpy(buf + sizeof(block_height), reinterpret_cast<void *>(&service_node_index), sizeof(service_node_index));
+      crypto::cn_fast_hash(buf, buf_size, result);
+    }
+    else
+    {
+      constexpr int buf_size = sizeof(block_height) + sizeof(service_node_index) + sizeof(vote_type) + sizeof(version);
+      char buf[buf_size];
+      char *buf_ptr = buf;
+
+      memcpy(buf_ptr, reinterpret_cast<void *>(&block_height),       sizeof(block_height));       buf_ptr += sizeof(block_height);
+      memcpy(buf_ptr, reinterpret_cast<void *>(&service_node_index), sizeof(service_node_index)); buf_ptr += sizeof(service_node_index);
+      memcpy(buf_ptr, reinterpret_cast<void *>(&vote_type),          sizeof(vote_type));          buf_ptr += sizeof(vote_type);
+      memcpy(buf_ptr, reinterpret_cast<void *>(&version),            sizeof(version));            buf_ptr += sizeof(version);
```

Comment:
These `reinterpret_cast<void *>` aren't necessary at all -- any pointer in C++ is required to be implicitly castable to a `void *`.

They also need to be endian-converted.

But also, looking at this I'm thinking I should pull this commit back in: https://github.com/loki-project/loki/pull/655/commits/e81350cb39ffecf314d03c873b72ba7ca6767a6c

It would let you replace this whole thing (including the endian conversion) with:

```C++
auto buf = tools::memcpy_le(block_height, service_node_index, vote_type, version);
```
without worrying about the size calculations.